### PR TITLE
sl-51: Modal component

### PR DESCRIPTION
**Pull request by jaroslawtt**
What should I implement else in this task?

Now it looks like: 
modal_proto

### CODE REVIEW

**Review by what1s1ove — CHANGES_REQUESTED** [review]
you need to implement background drop using ReactPortal

**Comment by jaroslawtt** [comment]
> you need to implement background drop using ReactPortal

image

File: frontend/src/libs/components/modal-window/modal-window.tsx [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import styles from './styles.module.scss';
+
+const ModalWindow: React.FC = () => {
+  return (
+    <Portal>
+      <div className={styles.backdrop} />
+      <div className={styles.modal}>
+        <button className={styles.modalPopUpClose}>
+          <img alt="close" src={closePopUpIcon} />
+        </button>
```
Comment (andanf-e) [inline frontend/src/libs/components/modal-window/modal-window.tsx]: We already have `IconButton` component - enhance Icons map with `close` svg icon and use `IconButton`

**Review by andanf-e — CHANGES_REQUESTED** [review]

File: frontend/src/libs/components/modal-window/modal-window.tsx [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { Portal } from 'react-portal';
```
Comment (andanf-e) [inline frontend/src/libs/components/modal-window/modal-window.tsx]: This package makes nothing special against `createPortal` from `react-dom`
```suggestion

```

**Review by andanf-e — CHANGES_REQUESTED** [review]

File: frontend/src/libs/components/components.ts [hunk 3]
```diff
@@ -6,6 +6,7 @@ export { IconButton } from './icon-button/icon-button.js';
 export { Input } from './input/input.js';
 export { Link } from './link/link.js';
 export { Loader } from './loader/loader.js';
+export { ModalWindow } from './modal-window/modal-window.js';
```
Comment (what1s1ove) [inline frontend/src/libs/components/components.ts]:
```suggestion
export { Modal } from './modal/modal.js';
```

File: frontend/src/libs/components/icon/common.ts [hunk 4]
```diff
@@ -8,6 +9,7 @@ const iconNameToSvgIcon: Record<
 > = {
   'arrowRight': ArrowRight,
   'arrowLeft': ArrowLeft,
+  'closePopUp': ClosePopUp,
```
Comment (what1s1ove) [inline frontend/src/libs/components/icon/common.ts]:
```suggestion
  'cross': Cross,
```

File: frontend/src/libs/components/icon/common.ts [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import { ReactComponent as ArrowLeft } from '~/assets/img/arrow-left.svg';
 import { ReactComponent as ArrowRight } from '~/assets/img/arrow-right.svg';
+import { ReactComponent as ClosePopUp } from '~/assets/img/close-pop-up-icon.svg';
```
Comment (what1s1ove) [inline frontend/src/libs/components/icon/common.ts]:
```suggestion
import { ReactComponent as ClosePopUp } from '~/assets/img/cross.svg';
```

File: frontend/src/libs/components/modal-window/modal-window.tsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import ReactDOM from 'react-dom';
+
+import { IconButton } from '~/libs/components/components.js';
+import { getValidClassNames } from '~/libs/helpers/helpers.js';
+
+import styles from './styles.module.scss';
+
+const ModalWindow: React.FC = () => {
```
Comment (what1s1ove) [inline frontend/src/libs/components/modal-window/modal-window.tsx]: don't forget about children in props

File: frontend/src/libs/components/modal-window/modal-window.tsx [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import ReactDOM from 'react-dom';
```
Comment (what1s1ove) [inline frontend/src/libs/components/modal-window/modal-window.tsx]:
```suggestion
import { createPortal } from 'react-dom';
```

File: frontend/src/libs/components/modal-window/modal-window.tsx [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+import { IconButton } from '~/libs/components/components.js';
+import { getValidClassNames } from '~/libs/helpers/helpers.js';
+
+import styles from './styles.module.scss';
+
+const ModalWindow: React.FC = () => {
+  return ReactDOM.createPortal(
+    <>
```
Comment (what1s1ove) [inline frontend/src/libs/components/modal-window/modal-window.tsx]: use https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog for modal

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/libs/components/modal-window/modal-window.tsx [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+      <div className={styles.backdrop} />
+      <div className={styles.modal}>
+        <IconButton
```
Comment (what1s1ove) [inline frontend/src/libs/components/modal-window/modal-window.tsx]: add cb `onClose` for it

File: frontend/src/libs/components/icon/common.ts (hunk 5 above)
Comment (andanf-e) [reply frontend/src/libs/components/icon/common.ts]: *cross

File: frontend/src/libs/components/modal-window/modal-window.tsx (hunk 8 above)
Comment (what1s1ove) [inline frontend/src/libs/components/modal-window/modal-window.tsx]: add `isOpen` state for it

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/libs/components/modal-window/modal-window.tsx [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+          icon="closePopUp"
+          label="Close modal"
+          className={getValidClassNames('', styles.modalPopUpClose)}
```
Comment (what1s1ove) [inline frontend/src/libs/components/modal-window/modal-window.tsx]:
```suggestion
          className={styles.modalPopUpClose}
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/libs/components/modal-window/modal-window.tsx (hunk 10 above)
Comment (jaroslawtt) [reply frontend/src/libs/components/modal-window/modal-window.tsx]: In this case I get lint:type error 
image

Comment (jaroslawtt) [reply frontend/src/libs/components/modal-window/modal-window.tsx]: Should I add like "as string"?

Comment (what1s1ove) [reply frontend/src/libs/components/modal-window/modal-window.tsx]: `styles.modalPopUpClose as string`

File: frontend/src/libs/components/modal/modal.tsx [hunk 11]
```diff
@@ -0,0 +1,42 @@
+import { useCallback } from 'react';
```
Comment (what1s1ove) [inline frontend/src/libs/components/modal/modal.tsx]: should be from `~/libs/hooks`

File: frontend/src/libs/components/modal/modal.tsx [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+  const [isOpen, setIsOpen] = useState(true);
+
+  const onClose = useCallback(() => setIsOpen(false), [setIsOpen]);
+
+  return (
+    <>
+      {isOpen &&
+        createPortal(
+          <>
+            <dialog className={styles.modal} id="modal">
```
Comment (what1s1ove) [inline frontend/src/libs/components/modal/modal.tsx]:
```suggestion
const Modal: React.FC<Properties> = ({ children, isOpen, onClose }: Properties) => {

  if (!isOpen) {
    return null
  }

  return (
    <>
      {isOpen &&
        createPortal(
          <>
            <dialog className={styles.modal} open={isOpen}>
```

File: frontend/src/libs/components/modal/modal.tsx [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+const Modal: React.FC<Properties> = ({ children }: Properties) => {
+  const [isOpen, setIsOpen] = useState(true);
+
+  const onClose = useCallback(() => setIsOpen(false), [setIsOpen]);
+
+  return (
+    <>
+      {isOpen &&
+        createPortal(
+          <>
```
Comment (what1s1ove) [inline frontend/src/libs/components/modal/modal.tsx]:
```suggestion
```
not sure that you need this fragment

File: frontend/src/libs/components/modal/modal.tsx [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+              <div className={styles.modalInner}>
+                <IconButton
+                  icon="cross"
+                  label="Close modal"
+                  className={styles.modalPopUpClose as string}
+                  onClick={onClose}
+                />
+                {children}
+              </div>
+              <div className={styles.backdrop} />
```
Comment (what1s1ove) [inline frontend/src/libs/components/modal/modal.tsx]:
```suggestion
```
https://developer.mozilla.org/en-US/docs/Web/CSS/::backdrop

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/libs/components/modal/modal.tsx (hunk 14 above)
Comment (jaroslawtt) [reply frontend/src/libs/components/modal/modal.tsx]: dialog element has by default set value background-color: canvas and it overwrites my styles used with backdrop pseudo-element. So should I set backdrop as backdround styles for modal class? 
image

Comment (jaroslawtt) [reply frontend/src/libs/components/modal/modal.tsx]:
```
.modal {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  border: none;
  background: color(srgb 0 0 0 / 60%);
}
```

Comment (what1s1ove) [reply frontend/src/libs/components/modal/modal.tsx]: you need to use `::backdrop` as the modal background

File: frontend/src/libs/components/modal/styles.module.scss [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+  width: 0;
+}
+
+.modal::backdrop {
+  background: color(srgb 0 0 0 / 60%);
+}
+
+.modal-inner {
+  position: relative;
+  z-index: 10;
```
Comment (andanf-e) [inline frontend/src/libs/components/modal/styles.module.scss]: Let's have `z-index` values defined in variables?

File: frontend/src/libs/components/modal/styles.module.scss [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+  width: 940px;
+  height: 665px;
+  background-color: var(--white);
+  border: none;
+}
+
+.modal-pop-up-close {
+  position: absolute;
```
Comment (andanf-e) [inline frontend/src/libs/components/modal/styles.module.scss]: Maybe let's better add header block to modal and just place close btn at the end of block?

File: frontend/src/assets/css/variables.scss [hunk 17]
```diff
@@ -12,4 +12,5 @@
   --blue: #3b79c3;
   --text-type-400: #130138;
   --text-type-300: #5a6474;
+  --z-index-modal: 10;
```
Comment (andanf-e) [inline frontend/src/assets/css/variables.scss]: As not it's global variable - then we don't need to name it as related to modal
```suggestion
  --z-index-level-one: 10;
```

File: frontend/src/libs/components/modal/modal.tsx [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+import { IconButton } from '~/libs/components/components.js';
+import { useEffect, useRef } from '~/libs/hooks/hooks.js';
+
+import styles from './styles.module.scss';
+
+type Properties = {
+  isOpen: boolean;
+  onClose: () => void;
+} & React.PropsWithChildren;
```
Comment (what1s1ove) [inline frontend/src/libs/components/modal/modal.tsx]:
```suggestion
};
```
describe children

File: frontend/src/libs/components/modal/modal.tsx [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+  useEffect(() => {
+    if (!reference.current) {
+      return;
+    }
+
+    if (isOpen && !reference.current.hasAttribute('open')) {
+      return reference.current.showModal();
+    } else if (!isOpen && reference.current.hasAttribute('open')) {
+      return reference.current.close();
+    }
```
Comment (andanf-e) [inline frontend/src/libs/components/modal/modal.tsx]: Do we really need this?

Comment (jaroslawtt) [reply frontend/src/libs/components/modal/modal.tsx]: If remove it in the DOM we have like this after modal is closed
image
But if we keep logic of setting dialog close it is clear.

Comment (jaroslawtt) [reply frontend/src/libs/components/modal/modal.tsx]: Or not, my fault, anyway have it

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/libs/components/modal/modal.tsx (hunk 18 above)
Comment (jaroslawtt) [reply frontend/src/libs/components/modal/modal.tsx]: should they be optional or not?

**Review by andanf-e — APPROVED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**